File: frontend/src/services/assignment.ts
import type { AssignmentScore, Vehicle, VehicleType } from '@/types';

/** Rough per-class cruising speed (km/h) used for ETA estimates. */
const CLASS_AVG_SPEED: Record<VehicleType, number> = {
  'heavy-truck': 45,
  'mini-truck': 40,
  tempo: 35,
  'delivery-van': 32,
};

/** CO₂ per litre of fuel burned (kg) — diesel-centric approximation. */
const CO2_PER_LITRE = 2.68;

export interface RankVehiclesInput {
  vehicles: Vehicle[];
  /** Shipment cargo weight in kg (0 = capacity-agnostic). */
  weightKg: number;
  /** Expected route distance in km. */
  distanceKm: number;
}

/**
 * Phase 5 — Adaptive Vehicle Assignment (client-side scorer).
 *
 * Scores every vehicle in the *user's own fleet* against a shipment load.
 * The score function is deliberately transparent: cost-efficiency dominates,
 * then fuel economy, fleet health, availability and capacity fit — and every
 * decision carries a human-readable reason, so the recommendation explains
 * itself (e.g. "Mini Truck selected because: Capacity matches requirement,
 * Lower fuel cost, Lowest estimated delivery cost").
 */
export function rankVehiclesForLoad({ vehicles, weightKg, distanceKm }: RankVehiclesInput): AssignmentScore[] {
  if (vehicles.length === 0) return [];
  const dist = Math.max(distanceKm, 0);
  const weight = Math.max(weightKg, 0);

  const maxCost = Math.max(...vehicles.map((v) => v.operatingCostPerKm));
  const minCost = Math.min(...vehicles.map((v) => v.operatingCostPerKm));
  const costRange = Math.max(maxCost - minCost, 1e-6);
  const maxMileage = Math.max(...vehicles.map((v) => Math.max(v.mileageKmpl, 0.1)));

  const scored = vehicles.map((v) => {
    const capacityOkay = weight === 0 || v.capacityKg >= weight;
    const etaMin = dist / CLASS_AVG_SPEED[v.type] * 60;

    const reasons: string[] = [];
    if (!v.available) reasons.push('Vehicle is not currently available');
    if (!capacityOkay) {
      reasons.push(
        `Capacity ${v.capacityKg.toLocaleString()} kg below shipment weight ${weight.toLocaleString()} kg`,
      );
    }
    if (capacityOkay) reasons.push('Capacity matches requirement');
    if (v.available) reasons.push('Vehicle available now');

    // Normalised sub-scores (0..1, higher is better).
    const costScore = 1 - (v.operatingCostPerKm - minCost) / costRange;
    const efficiencyScore = Math.max(v.mileageKmpl, 0.1) / maxMileage;
    const healthScore = v.healthScore / 100;
    const availabilityScore = v.available ? 1 : 0;
    const capacityFit = capacityOkay ? 1 : 0;

    let score =
      0.4 * costScore +
      0.2 * efficiencyScore +
      0.15 * healthScore +
      0.15 * availabilityScore +
      0.1 * capacityFit;

    if (!capacityOkay || !v.available) score *= 0.1;

    // Why-rationale, cost-driven phrasing used by the advisor copy.
    if (v.operatingCostPerKm === minCost) reasons.push('Lowest operating cost per km');
    else if (v.operatingCostPerKm <= maxCost * 0.55) reasons.push('Lower fuel cost');
    if (v.mileageKmpl === maxMileage) reasons.push('Best fuel efficiency in fleet');

    const remaining = Math.max(
      0,
      v.maintenanceIntervalKm - (v.currentOdometerKm - v.lastMaintenanceKm),
    );
    if (remaining <= 500) {
      reasons.push(`Service due within ${remaining.toLocaleString()} km — schedule maintenance`);
    }

    return {
      vehicleId: v.id,
      vehicleName: v.name,
      plateNumber: v.plateNumber,
      vehicleType: v.type,
      availability: v.available,
      capacityOk: capacityOkay,
      score: Math.round(score * 1000) / 1000,
      costEstimate: Math.round(dist * v.operatingCostPerKm),
      co2EstimateKg: Math.round((dist / Math.max(v.mileageKmpl, 0.1)) * CO2_PER_LITRE),
      etaMin: Math.round(etaMin),
      reasons: reasons.slice(0, 3),
    };
  });

  scored.sort((a, b) => b.score - a.score);
  return scored;
}

File: backend/app/services/assignment.py
"""Heuristic vehicle-assignment scoring.

Full AI assignment (ML-driven) lands in Phase 7; this deterministic scorer
already prepares the JSON contract and decision reasons used by the UI.
"""

from __future__ import annotations

from app.schemas.routing import (
    AssignPlanRequest,
    AssignmentScore,
    OptimizeResult,
    VehicleCandidate,
    VehicleParams,
)
from app.services.routing import compute_optimized_route, congestion_factor


def _vehicle_cost_estimate(vehicle: VehicleCandidate, distance_km: float) -> float:
    return distance_km * vehicle.operating_cost_per_km


def _vehicle_co2(vehicle: VehicleCandidate, distance_km: float) -> float:
    from app.core.config import settings

    mileage = max(vehicle.mileage_kmpl, 0.1)
    return (distance_km / mileage) * settings.co2_per_litre_fuel


async def plan_assignment(req: AssignPlanRequest) -> tuple[OptimizeResult, list[AssignmentScore]]:
    """Score candidate vehicles against a route and return a ranked list."""
    route = await compute_optimized_route(
        OptimizeRequest(
            start=req.start,
            stops=req.stops,
            destination=req.destination,
            profile=req.profile,
            vehicle=VehicleParams(),
            weight_kg=req.weight_kg,
        )
    )

    dist = route.distance_km
    factor, _ = congestion_factor()
    eta = route.duration_min

    scored: list[AssignmentScore] = []

    for v in req.vehicles:
        capacity_ok = v.capacity_kg >= req.weight_kg or req.weight_kg <= 0
        cost = _vehicle_cost_estimate(v, dist)
        co2 = _vehicle_co2(v, dist)

        reasons: list[str] = []
        if not v.available:
            reasons.append("Vehicle is not currently available")
        if not capacity_ok:
            reasons.append(f"Capacity {v.capacity_kg} kg below shipment weight {int(req.weight_kg)} kg")

        # Normalised sub-scores (0..1, higher is better).
        # Cost dominates, then ETA, then fleet health, then availability.
        max_cost = max((x.operating_cost_per_km for x in req.vehicles), default=1)
        min_cost = min((x.operating_cost_per_km for x in req.vehicles), default=1)
        cost_range = max(max_cost - min_cost, 1e-6)
        cost_score = 1 - ((v.operating_cost_per_km - min_cost) / cost_range)

        eta_score = max(0.0, 1 - (eta / (eta + 60)))
        health_score = v.health_score / 100.0
        avail_score = 1.0 if v.available else 0.0

        score = 0.5 * cost_score + 0.25 * eta_score + 0.15 * health_score + 0.1 * avail_score
        if not capacity_ok or not v.available:
            score *= 0.1

        if capacity_ok:
            reasons.append("Capacity matches shipment load")
        if v.available:
            reasons.append("Vehicle available now")

        scored.append(
            AssignmentScore(
                vehicle_id=v.id,
                vehicle_name=v.name,
                plate_number=v.plate_number,
                vehicle_type=v.vehicle_type,
                availability=v.available,
                capacity_ok=capacity_ok,
                score=round(score, 3),
                cost_estimate=round(cost, 0),
                co2_estimate_kg=round(co2, 0),
                eta_min=round(eta, 1),
                reasons=reasons[:3],
            )
        )

    scored.sort(key=lambda s: s.score, reverse=True)
    return route, scored